### ISSUE TITLE
Increase in 'list nodes' latency on large cluster

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#64022]
We started seeing this in our 5k-node density test. This is **not** causing SLO violations or failures in the test - so it's not a release-blocking issue. But we should still take a look into it.

![list-nodes](https://user-images.githubusercontent.com/4333971/40229856-d0c6d7fc-5a95-11e8-9c48-3b54e80374c9.png)

Also, there's some increase we saw in the no. of list nodes calls around the same time (i.e b/w runs 142 and 145):

![count](https://user-images.githubusercontent.com/4333971/40230017-66a703e6-5a96-11e8-8038-7d1335d9368e.png)

cc @kubernetes/sig-scalability-bugs 
/assign @wojtek-t 
(who already started looking into it)

**Comment by wojtek-t** [comment]
I already debugged this issue.

With https://github.com/kubernetes/test-infra/pull/7824, we enabled running heapster and metric server in our scalability tests (for 2k and 5k node clusters).

After that change, the number of "LIST nodes" calls increased.
I looked deeper into the metrics and basically:
- up until run 142, there were 0 calls coming from pod-nanny
- from run 145, there is a number (in 145 - 360 of those) calls from pod-nanny

Pretty much all of these calls are really long, so they are clearly the reason for that regression.

That said, I looked even deeper into it. Interestingly pod-nanny is listing nodes from cache, so it's surprising why they are so long. 

So I looked into currently running test and the reason is the following;

- [BTW - there are 2 pod nannies, one for heapster and one for metrics-server]
- both pod nannies has CPU-limit set:
  - metric server: 100m cpu limit:
https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/metrics-server/metrics-server-deployment.yaml#L63
  - heapster: 50m cpu limit:
https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/cluster-monitoring/standalone/heapster-controller.yaml#L65

Looking into usage, both these container are fully saturated all the time.

So to summarize, the reason for regression is:
- pod-nannies have cpu-limits
- those limits are too low (for large clusters)
- they are cpu-starved
- as a result, they are not able to quickly process/read data that apiserver is sending - that's why the requests take so long

**Comment by wojtek-t** [comment]
Longer term: https://github.com/kubernetes/community/pull/2090 should significantly help.

Shorter-term, the only solutions that I see is:
- increasing the limit for pod-nanny
- or somehow autoscale it based on cluster size (so basically nanny-for-nanny concept)

**Comment by shyamjvs** [comment]
cc @kubernetes/sig-instrumentation-bugs @piosz

**Comment by shyamjvs** [comment]
Thanks for digging up @wojtek-t.

Btw - why do you call the 'setting proper resource limits for heapster' part short-term? Shouldn't we be doing that anyway or you mean it won't be needed once we have cheaper heartbeats?

**Comment by wojtek-t** [comment]
https://github.com/kubernetes/autoscaler/pull/1041 + https://github.com/kubernetes/kubernetes/pull/65864 should at least partially help

**Comment by wojtek-t** [comment]
We fixed that by list only object metadata.
